server/utils/calendarDay.ts

```diff
@@ -0,0 +1,38 @@
+export default class CalendarDay {
+  private constructor(readonly day: number, readonly month: number, readonly year: number) {}
+
+  static fromComponents(day: number, month: number, year: number): CalendarDay | null {
+    return isValidDate(day, month, year) ? new CalendarDay(day, month, year) : null
+  }
+
+  static parseIso8601(val: string): CalendarDay | null {
+    const result = /^(\d{4})-(\d{2})-(\d{2})$/.exec(val)
+
+    if (result === null) {
+      return null
+    }
+
+    const year = Number(result[1])
+    const month = Number(result[2])
+    const day = Number(result[3])
+
+    if (!isValidDate(day, month, year)) {
+      return null
+    }
+
+    return new CalendarDay(day, month, year)
+  }
+
+  get iso8601(): string {
+    return [
+      String(this.year).padStart(4, '0'),
+      String(this.month).padStart(2, '0'),
+      String(this.day).padStart(2, '0'),
+    ].join('-')
+  }
+}
+
+function isValidDate(day: number, month: number, year: number) {
+  const date = new Date(Date.UTC(year, month - 1, day))
```

Comment:
could do with comment explaining what's going on here (i assume we are checking for overflows)